src/pixelator/plot/__init__.py

```diff
@@ -44,6 +57,697 @@ def _calculate_densities(coordinates, distance_cutoff, unit_sphere_surface):
+    :param cache_layout: Whether to cache the layout coordinates. Defaults to False.
+    :param suppress_fig: Whether to suppress (i.e. not plot) the figure. Defaults to False.
+    :return: The plotted 3D graph.
+    :rtype: go.Figure
+
+    """
+    component_graph = _get_component_graph(pxl_data=pxl_data, component=component)
+    filtered_coordinates, _ = _get_coordinates(
+        component_graph=component_graph,
+        layout_algorithm=layout_algorithm,
+        cache_layout=cache_layout,
+        show_b_nodes=show_b_nodes,
+    )
+    if marker is not None and log_scale:
+        filtered_coordinates["color"] = np.log1p(filtered_coordinates[marker])
+    else:
+        filtered_coordinates["color"] = filtered_coordinates[marker]
+
+    if normalize:
+        filtered_coordinates[["x", "y", "z"]] = filtered_coordinates[
+            ["x_norm", "y_norm", "z_norm"]
+        ]
+
+    fig = plot_3d_from_coordinates(
+        coordinates=filtered_coordinates,
+        node_size=node_size,
+        opacity=opacity,
+        cmap=cmap,
+        suppress_fig=True,
+    )
+
+    fig.update_layout(title=component)
+
+    if not suppress_fig:
+        fig.show()
+
+    return fig
+
+
+def _pivot_colocalization_data(
+    colocalization_data: pd.DataFrame,
+    value_col: str = "pearson",
+    markers: Union[list, None] = None,
+):
+    colocalization_data_pivot = pd.pivot_table(
+        colocalization_data,
+        index="marker_1",
+        columns="marker_2",
+        values=value_col,
+        fill_value=0,
+    )
+
+    if markers is not None:
+        colocalization_data_pivot = colocalization_data_pivot.loc[markers, markers]
+
+    for m in colocalization_data_pivot.index:
+        colocalization_data_pivot.loc[m, m] = 0  # remove autocorrelations
+
+    colocalization_data_pivot = colocalization_data_pivot.fillna(0)
+
+    return colocalization_data_pivot
+
+
+def _make_colocalization_symmetric(
+    colocalization_data: pd.DataFrame,
+    value_col: str = "pearson",
+):
+    colocalization_data = pd.DataFrame(
+        np.concatenate(
+            [
+                colocalization_data[["marker_1", "marker_2", value_col]].to_numpy(),
+                colocalization_data[["marker_2", "marker_1", value_col]].to_numpy(),
+            ],
+        ),
+        columns=["marker_1", "marker_2", value_col],
+    )
+    colocalization_data = (
+        colocalization_data.groupby(["marker_1", "marker_2"])[value_col]
+        .apply(lambda x: np.mean(x))
+        .reset_index()
+    )
+
+    return colocalization_data
+
+
+def plot_colocalization_heatmap(
+    colocalization_data: pd.DataFrame,
+    markers: Union[list, None] = None,
+    cmap: str = "vlag",
+    use_z_scores: bool = False,
+) -> Tuple[plt.Figure, plt.Axes]:
+    """Plot a colocalization heatmap based on the provided colocalization data. The colocalization_data DataFrame should contain the columns "marker_1", "marker_2", "pearson", "pearson_z".
```

Comment:
The line length his is also to long, same on 627. Sorry to nit pick over this - I will look and see if I can figure out what the linter is not picking up on this.